### PR TITLE
Http2 v2

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/1947

DRAFT for discussion
Not to be merged

Describe changes:
- Adds basic HTTP2 parser

Much more work to be done :
 + Commit squashing
 + parse deeper
 + handle transactions with progress completion
 + more generic code for logging
 + connection upgrade from HTTP1
 + rules file for applayer events
 + tests
 + robustness against evasions

Modifies #4701 with 
- rebasing
- buffering using `AppLayerResult::incomplete`
- parsing/logging/detecting GOAWAY error code

### CODE REVIEW

File: rust/src/http2/http2.rs [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    //we may have consumed data from previous records
+                    if input.len() < HTTP2_FRAME_HEADER_LEN {
+                        return AppLayerResult::incomplete(
+                            (il - input.len()) as u32,
+                            HTTP2_FRAME_HEADER_LEN as u32,
+                        );
+                    } else {
+                        //TODO BUG_ON ?
```
Comment (catenacyber) [inline rust/src/http2/http2.rs]: Is there something equivalent to `BUG_ON` in Rust ? @jasonish

File: src/detect-http2.c [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    sigmatch_table[DETECT_HTTP2_ERRORCODE].name = "http2.errorcode";
+    sigmatch_table[DETECT_HTTP2_ERRORCODE].desc = "match on HTTP2 error code field";
+    sigmatch_table[DETECT_HTTP2_ERRORCODE].url = DOC_URL DOC_VERSION "/rules/http2-keywords.html#errorcode";
+    sigmatch_table[DETECT_HTTP2_ERRORCODE].Match = NULL;
+    sigmatch_table[DETECT_HTTP2_ERRORCODE].AppLayerTxMatch = DetectHTTP2errorcodeMatch;
+    sigmatch_table[DETECT_HTTP2_ERRORCODE].Setup = DetectHTTP2errorcodeSetup;
+    sigmatch_table[DETECT_HTTP2_ERRORCODE].Free = DetectHTTP2errorcodeFree;
+#ifdef UNITTESTS
+    //TODO should we call multiple times DetectHTTP2RegisterTests ?
+    sigmatch_table[DETECT_HTTP2_ERRORCODE].RegisterTests = DetectHTTP2RegisterTests;
```
Comment (catenacyber) [inline src/detect-http2.c]: I write TODOs to ask questions, so 

Should we call multiple times `DetectHTTP2RegisterTests` ? @victorjulien

File: src/detect-http2.c [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+{
+    uint32_t *detect = (uint32_t *)ctx;
+
+    int value = rs_http2_tx_get_errorcode(txv, flags);
+    if (value < 0) {
+        //no value, no match
+        return 0;
+    }
+    return *detect == (uint32_t) value;
```
Comment (catenacyber) [inline src/detect-http2.c]: I guess we should improve this...
Any thoughts ?

File: rust/src/http2/http2.rs [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+                    let mut tx = self.new_tx();
+                    tx.ftype = Some(head.ftype);
+                    match head.ftype {
+                        parser::HTTP2FrameType::GOAWAY => {
+                            match parser::http2_parse_frame_goaway(rem) {
+                                Ok((_, goaway)) => {
+                                    tx.type_data = Some(HTTP2FrameTypeData::GOAWAY(goaway));
+                                }
+                                Err(nom::Err::Incomplete(_)) => {
+                                    return AppLayerResult::incomplete(
```
Comment (catenacyber) [inline rust/src/http2/http2.rs]: This seems the right way to do so far :
- parse frame header
- keep state of remaining frame bytes
- use tcp buffering if we need to parse deeper

Should be `HTTP2_FRAME_HEADER_LEN + 4`

File: src/detect-http2.c (hunk 2 above)
Comment (victorjulien) [reply src/detect-http2.c:89]: Once. Otherwise it would register the same tests multiple times.

File: rust/src/http2/http2.rs (hunk 1 above)
Comment (victorjulien) [reply rust/src/http2/http2.rs:230]: `panic!("msg");`

File: src/detect-http2.c (hunk 3 above)
Comment (victorjulien) [reply src/detect-http2.c:212]: Seems like premature optimization to me.

File: rust/src/http2/http2.rs (hunk 4 above)
Comment (victorjulien) [reply rust/src/http2/http2.rs:198]: `nom::Err::Incomplete(_)` should tell you how many more bytes it needs. Can you use that too?

File: src/detect-http2.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-http2.c:89]: ok thanks

File: src/detect-http2.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-http2.c:212]: So, just leave a TODO comment for now ?

Comment (victorjulien) [reply src/detect-http2.c:212]: It's likely we'll want things like negation or ranges here too

Comment (catenacyber) [reply src/detect-http2.c:212]: Ok for negation, but not sure for ranges as this field is an enumeration

File: rust/src/http2/http2.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/http2/http2.rs:230]: ok thanks

File: src/detect-http2.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-http2.c:212]: And do not we have generic negation for whatever keyword (returning the negation of the Match function) ?

File: rust/src/http2/http2.rs (hunk 4 above)
Comment (catenacyber) [reply rust/src/http2/http2.rs:198]: Ok.
Should we try to be smarter : when we have a composite type, u16 and u32 for instance, we know we will need 6 bytes, even if nom tells us we first need 2 bytes ?

**Pull request closed without merging**